File: ios/Stockbook/Features/Customers/StatementScreen.swift
import SwiftUI

/// What a delivery reads as on a statement, on screen and in the text copy alike.
///
/// A supplier's bill entered as a figure names nothing, and "× 0" of nothing is
/// not a line anybody can read — so it says what the paper is instead.
///
/// The describing itself is `Purchase.described`, on the model, where the Kotlin
/// twin has the identical property: a statement is a document the two apps must
/// word the same way.
@MainActor
private func deliveryDetail(_ purchase: Purchase) -> String {
    purchase.described ?? Loc.supplierBillTitle
}

/// One customer's account over a period, as a document.
///
/// Full screen rather than a sheet, for two reasons: it can run to a page, and it
/// is the one thing in this app the owner may well turn round and show the person
/// it is about. Everything on it is drawn from `Statement`, which does the
/// arithmetic and is tested against literal figures.
struct StatementScreen: View {
    @Environment(StockbookStore.self) private var store
    @Environment(AppRouter.self) private var router
    @Environment(\.currency) private var currency

    /// Set once the PDF is on disk; the share sheet is presented off it.
    @State private var pdfFile: StatementFile?

    /// Whose account: a customer key, or a supplier key with `isSupplier` set.
    let partyKey: String

    /// Which side of the book. One screen for both, because a statement is a
    /// statement — see `Statement.make`, where the same arithmetic serves both —
    /// and two screens would drift the moment either was corrected.
    var isSupplier = false
    let onClose: () -> Void

    /// Which chip is on.
    ///
    @State private var choice: PeriodChoice = .thisMonth

    @State private var from = Calendar.current.date(byAdding: .month, value: -1, to: .now) ?? .now
    @State private var to = Date.now

    private var period: StatementPeriod { choice.period(from: from, to: to) }

    private var statement: Statement? {
        isSupplier
            ? store.statementForSupplier(key: partyKey, period: period)
            : store.statement(forCustomer: partyKey, period: period)
    }

    var body: some View {
        VStack(spacing: 0) {
            ScreenHeader(title: Loc.statement, subtitle: statement?.party.name) {
                Button(Loc.done, action: onClose)
                    .buttonStyle(GhostButtonStyle(fontSize: 12.5))
            }

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    PeriodPicker(choice: $choice, from: $from, to: $to).padding(.bottom, 10)

                    if let statement {
                        contactLine(statement.party)
                        document(statement)
                        shareRow(statement)
                    }
                }
                .padding(.horizontal, Metrics.screenPadding)
                .padding(.bottom, 24)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
        .background(Nocturne.bg.ignoresSafeArea())
        .motion(Motion.list, value: choice)
        // Presented off the file rather than a flag, so the sheet cannot open
        // before the PDF exists — and closing it drops the URL, so the next tap
        // renders the period the owner is looking at *then*.
        .sheet(item: $pdfFile) { file in
            ShareSheet(url: file.url)
                .presentationDetents([.medium, .large])
        }
    }

    // MARK: The document

    @ViewBuilder
    private func contactLine(_ party: StatementParty) -> some View {
        let details = [party.phone, party.place].compactMap { $0 }
        if !details.isEmpty {
            Text(details.joined(separator: " · "))
                .nocturneText(.meta)
                .padding(.bottom, 8)
        }
    }

    private func document(_ statement: Statement) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Kicker(Loc.dateSpan(from: Loc.longDate(statement.range.start), to: Loc.longDate(lastDay(of: statement.range))))
                .padding(.bottom, 12)

            row(Loc.openingBalance, Money.text(statement.openingBalance, in: currency), muted: true)

            FadedRule().padding(.vertical, 10)

            if statement.isEmpty {
                Text(Loc.nothingInThisPeriod)
                    .nocturneText(.meta)
                    .frame(maxWidth: .infinity, alignment: .center)
                    .padding(.vertical, 14)
            } else {
                VStack(alignment: .leading, spacing: 10) {
                    ForEach(Array(statement.entries.enumerated()), id: \.element.id) { index, entry in
                        entryRow(entry, balance: statement.runningBalances[index])
                    }
                }
            }

            FadedRule().padding(.vertical, 10)

            row(chargedLabel(statement), Money.text(statement.billed, in: currency))
            row(settledLabel(statement), Money.text(statement.received, in: currency))
            // Its own line, and only where there is one. Credit is not cash, and
            // a row saying "0.00 credited" on every statement teaches people to
            // stop reading the ones that are not zero.
            if statement.credited > 0 {
                row(Loc.creditNotes, Money.text(statement.credited, in: currency))
            }
            // Their own lines, and only where there is one — for the reason the
            // credit notes have theirs: a transfer in is not something the shop
            // invoiced and a transfer out is not money it took.
            if statement.transferredIn > 0 {
                row(Loc.transferredInLabel, Money.text(statement.transferredIn, in: currency))
            }
            if statement.transferredOut > 0 {
                row(Loc.transferredOutLabel, Money.text(statement.transferredOut, in: currency))
            }

            FadedRule().padding(.vertical, 10)

            // The number the whole document exists to state.
            HStack(alignment: .firstTextBaseline) {
                Text(Loc.closingBalance)
                    .font(NocturneType.inter(13, .medium))
                    .foregroundStyle(Nocturne.text)
                Spacer(minLength: 10)
                Text(closingText(statement))
                    .nocturneText(.bigNumber(22))
                    .foregroundStyle(statement.closingBalance > 0 ? Nocturne.accent : Nocturne.text)
                    .contentTransition(.numericText())
            }
            .motion(Motion.numbers, value: statement.closingBalance)
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.statRadius, style: .continuous))
        .hairline(radius: Metrics.statRadius)
    }

    private func closingText(_ statement: Statement) -> String {
        if statement.closingBalance < 0 {
            return Loc.inAdvance(Money.text(-statement.closingBalance, in: currency))
        }
        return Money.text(statement.closingBalance, in: currency)
    }

    private func row(_ label: String, _ value: String, muted: Bool = false) -> some View {
        HStack(alignment: .firstTextBaseline) {
            Text(label).nocturneText(muted ? .meta : .body)
            Spacer(minLength: 10)
            Text(value)
                .font(NocturneType.inter(13))
                .foregroundStyle(muted ? Nocturne.neutral500 : Nocturne.text)
        }
    }

    @ViewBuilder
    private func entryRow(_ entry: Statement.Entry, balance: Double) -> some View {
        entryLine(entry, balance: balance)
        .contentShape(Rectangle())
        .onTapGesture {
            // Both of the things this screen can correct open the same way: tap
            // the row, get the sheet it was written on, with removal one button
            // inside it. They used to differ — a credit note opened, a payment
            // armed a delete — which meant the same gesture did two things
            // depending on which row it landed on.
            switch entry {
            case .creditNote(let note):
                guard let customer = store.customer(key: note.customerKey) else { return }
                router.creditNoteFor = CreditNoteTarget(customer: customer, note: note)
            case .payment(let payment):
                // The customer comes with it: the sheet shows what will still be
                // owed once the correction is saved, which needs the whole
                // account rather than the one payment.
                guard let customer = store.customer(key: payment.customerKey) else { return }
                router.editingPayment = payment
                router.paymentFor = customer
            case .supplierPayment(let payment):
                guard let supplier = store.supplier(key: payment.supplierKey) else { return }
                router.editingSupplierPayment = payment
                router.supplierPaymentFor = supplier
            case .transfer:
                // Nothing to reopen. A transfer is an amount and a reason, and
                // correcting it means removing it and moving the right figure —
                // which the account screen it came from already offers.
                break
            case .bill, .purchase:
                // Not corrected from here. Removing one puts stock back on the
                // shelf, and offering that from a row on a document somebody is
                // reading would be a second, worse route than the opened document.
                break
            }
        }
    }

    @ViewBuilder
    private func entryLine(_ entry: Statement.Entry, balance: Double) -> some View {
        HStack(alignment: .top, spacing: 10) {
            VStack(alignment: .leading, spacing: 2) {
                switch entry {
                case .bill(let bill):
                    Text(reference(entry))
                        .font(NocturneType.inter(13))
                    Text(bill.summary)
                        .nocturneText(.meta)
                        .lineLimit(2)
                case .payment(let payment):
                    HStack(spacing: 5) {
                        Glyph(Icon.confirm, size: 10).foregroundStyle(Nocturne.accent400)
                        Text(reference(entry))
                            .font(NocturneType.inter(13))
                            .foregroundStyle(Nocturne.accent400)
                    }
                    if let note = payment.note {
                        Text(note).nocturneText(.meta)
                    }
                case .creditNote(let note):
                    Text(reference(entry))
                        .font(NocturneType.inter(13))
                        .foregroundStyle(Nocturne.accent400)
                    // Why it was written, where the owner said. On a document
                    // somebody is checking against their own paper, "returned,
                    // damaged" is the difference between a figure they
                    // recognise and one they have to go and ask about.
                    if let reason = note.reason {
                        Text(reason).nocturneText(.meta).lineLimit(2)
                    }
                case .purchase(let purchase):
                    Text(reference(entry))
                        .font(NocturneType.inter(13))
                    // The product and how many of it: a delivery note's whole
                    // content, on one line, since a purchase carries one product.
                    Text(deliveryDetail(purchase))
                        .nocturneText(.meta)
                        .lineLimit(2)
                case .transfer(let transfer, _, _):
                    // Named by the account at the other end — see `reference` —
                    // because there is no number to show. The reason is drawn
                    // under it for the same purpose a credit note's is: a figure
                    // the customer cannot place is one they come and ask about.
                    Text(reference(entry))
                        .font(NocturneType.inter(13))
                    if let note = transfer.note {
                        Text(note).nocturneText(.meta).lineLimit(2)
                    }
                case .supplierPayment(let payment):
                    HStack(spacing: 5) {
                        Glyph(Icon.confirm, size: 10).foregroundStyle(Nocturne.accent400)
                        Text(reference(entry))
                            .font(NocturneType.inter(13))
                            .foregroundStyle(Nocturne.accent400)
                    }
                    if let note = payment.note {
                        Text(note).nocturneText(.meta)
                    }
                }
                Text(Loc.longDate(entry.date)).nocturneText(.meta)
            }

            Spacer(minLength: 6)

            VStack(alignment: .trailing, spacing: 2) {
                Text(amountText(entry))
                    .font(NocturneType.inter(13))
                    .foregroundStyle(amountTint(entry))
                // The running balance beside every line: the column that turns a
                // list into a statement somebody can check.
                Text(Money.text(balance, in: currency))
                    .nocturneText(.meta)
            }
        }
    }

    /// "Billed" and "Received" are the customer's words. On a supplier's account
    /// they read backwards — the shop is the one being billed — so the two
    /// figures are named by which way the account runs.
    private func chargedLabel(_ statement: Statement) -> String {
        statement.party.isSupplier ? Loc.purchasedInPeriod : Loc.billedInPeriod
    }

    private func settledLabel(_ statement: Statement) -> String {
        statement.party.isSupplier ? Loc.paidOutInPeriod : Loc.receivedInPeriod
    }

    private func amountText(_ entry: Statement.Entry) -> String {
        switch entry {
        case .bill(let bill): Money.text(bill.total, in: currency)
        case .purchase(let purchase): Money.text(purchase.total, in: currency)
        // A minus sign on both kinds of payment: it is what the account moves by,
        // and on a supplier's statement that is money leaving rather than arriving.
        case .creditNote(let note): "− \(Money.text(note.total, in: currency))"
        case .payment(let payment): "− \(Money.text(payment.amount, in: currency))"
        case .supplierPayment(let payment): "− \(Money.text(payment.amount, in: currency))"
        // The sign says which way this account moved, not which way the money
        // went — no money went anywhere. Out of this account is a reduction like
        // a payment; into it is a charge like a bill.
        case .transfer(let transfer, let outgoing, _):
            outgoing
                ? "− \(Money.text(transfer.amount, in: currency))"
                : Money.text(transfer.amount, in: currency)
        }
    }

    private func amountTint(_ entry: Statement.Entry) -> Color {
        switch entry {
        case .bill, .purchase: Nocturne.text
        case .payment, .supplierPayment, .creditNote: Nocturne.accent400
        // Whichever way this one moves the account, so it reads like the charge
        // or the settlement it is.
        case .transfer(_, let outgoing, _): outgoing ? Nocturne.accent400 : Nocturne.text
        }
    }

    /// What a row is called on screen — the printed statement's own rule,
    /// borrowed rather than restated.
    private func reference(_ entry: Statement.Entry) -> String {
        StatementDocument.reference(entry, Loc)
    }

    /// The last day this statement can honestly say it covers — the range's own
    /// rule, so the screen and the PDF are headed with the same date.
    private func lastDay(of range: StatementRange) -> Date {
        range.asOf()
    }

    // MARK: Sharing

    /// One way out, and it is the document. The app makes no network call; the OS
    /// does whatever the owner picks.
    ///
    /// There was a plain-text share beside this for a quick message, and it was a
    /// second rendering of the same figures with none of the page's wording,
    /// arithmetic or letterhead — a statement somebody could quote back that the
    /// app would not recognise as its own.
    private func shareRow(_ statement: Statement) -> some View {
        Button(Loc.sharePdf) { makePDF(statement) }
            .buttonStyle(PrimaryButtonStyle(fullWidth: true, height: 44, fontSize: 13.5))
            .padding(.top, 10)
    }

    /// Renders the document to a file and hands the URL to the share sheet.
    ///
    /// A failure leaves `pdfFile` nil and nothing opens, which is the honest
    /// outcome: there is no half-written statement worth offering, and the
    /// figures are still on screen.
    private func makePDF(_ statement: Statement) {
        let document = StatementDocument.make(
            statement: statement,
            settings: store.settings,
            strings: Loc
        )
        let slug = document.partyName
            .replacingOccurrences(of: "[^A-Za-z0-9]+", with: "-", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "-"))
            .lowercased()
        guard let url = try? StatementPDF.write(
            document,
            fileName: Loc.statementFileName(name: slug, date: Copy.fileDate(.now))
        ) else { return }
        pdfFile = StatementFile(url: url)
    }
}

File: android/app/src/main/kotlin/com/stockbook/app/feature/customers/StatementScreen.kt
package com.stockbook.app.feature.customers

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.FadedRule
import com.stockbook.app.design.GhostButton
import com.stockbook.app.design.Glyph
import com.stockbook.app.design.Icon
import com.stockbook.app.design.Kicker
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.PeriodChoice
import com.stockbook.app.design.PeriodPicker
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.ScreenHeader
import com.stockbook.app.design.PrimaryButton
import com.stockbook.app.design.card
import com.stockbook.app.design.hairline
import com.stockbook.core.model.Currency
import com.stockbook.core.model.Statement
import com.stockbook.core.model.StatementParty
import com.stockbook.core.model.StatementPeriod
import com.stockbook.core.model.StatementRange
import com.stockbook.core.model.Timestamps
import com.stockbook.core.money.Money
import com.stockbook.core.store.StockbookStore
import com.stockbook.core.text.StatementDocument
import com.stockbook.core.text.Strings
import java.time.Instant
import java.time.ZoneId

/**
 * One customer's account over a period, as a document.
 *
 * Full screen rather than a sheet, for two reasons: it can run to a page, and it
 * is the one screen here the owner may well turn round and show the person it is
 * about. Everything on it comes from [Statement], which does the arithmetic and is
 * tested against literal figures.
 */
@Composable
fun StatementScreen(
    /** Whose account: a customer key, or a supplier key with [isSupplier] set. */
    partyKey: String,
    /**
     * Which side of the book. One screen for both, because a statement is a
     * statement — see `Statement.make`, where the same arithmetic serves both —
     * and two screens would drift the moment either was corrected.
     */
    isSupplier: Boolean = false,
    store: StockbookStore,
    currency: Currency,
    strings: Strings,
    /** Hands the rendered document to the share sheet. */
    onSharePdf: (StatementDocument) -> Unit = {},
    /**
     * Opens a credit note for correction. A note is edited from here because
     * this is the document it appears on — the same place a bill is opened from.
     */
    onEditCreditNote: (com.stockbook.core.model.CreditNote) -> Unit = {},
    /**
     * Opens a payment for correction, by id. Same reasoning as the credit note
     * above it: this is the document the payment appears on.
     */
    onEditPayment: (String) -> Unit = {},
    onClose: () -> Unit
) {
    /**
     * Which chip is on.
     *
     * Held as the *choice* rather than as a `StatementPeriod`, because a period
     * carries an instant: one built for the chip would never equal the one built a
     * moment earlier and stored, so no chip would ever look selected. The period
     * is derived from this instead, which also means "this month" is still this
     * month if the app is left open past midnight on the 1st.
     */
    var choice by remember { mutableStateOf(PeriodChoice.THIS_MONTH) }
    var from by remember {
        mutableStateOf(
            Timestamps.now().atZone(ZoneId.systemDefault()).toLocalDate()
                .minusMonths(1).atStartOfDay(ZoneId.systemDefault()).toInstant()
        )
    }
    var to by remember { mutableStateOf(Timestamps.now()) }

    val period = choice.period(from, to)

    val statement = if (isSupplier) {
        store.statementForSupplier(partyKey, period)
    } else {
        store.statementForCustomer(partyKey, period)
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Nocturne.bg)
            .statusBarsPadding()
    ) {
        ScreenHeader(title = strings.statement, subtitle = statement?.party?.name) {
            GhostButton(strings.done, onClick = onClose, fontSize = 12.5)
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = Metrics.screenPadding)
                .navigationBarsPadding()
                .padding(bottom = 24.dp)
        ) {
            PeriodPicker(
                choice = choice,
                from = from,
                to = to,
                strings = strings,
                onChoose = { choice = it },
                onFrom = { from = it },
                onTo = { to = it }
            )

            if (statement != null) {
                Spacer(Modifier.height(10.dp))
                ContactLine(statement.party)
                Document(
                    statement = statement,
                    currency = currency,
                    strings = strings,
                    onEditPayment = onEditPayment,
                    onEditCreditNote = onEditCreditNote
                )
                Spacer(Modifier.height(10.dp))
                // One way out, and it is the document. There was a plain-text
                // share beside this for a quick message, and it was a second
                // rendering of the same figures with none of the page's wording,
                // arithmetic or letterhead — a statement somebody could quote
                // back that the app would not recognise as its own.
                PrimaryButton(
                    strings.sharePdf,
                    onClick = { onSharePdf(StatementDocument.make(statement, store.settings, strings, currency)) },
                    fullWidth = true,
                    height = 44.dp,
                    fontSize = 13.5
                )
            }
        }
    }
}

@Composable
private fun ContactLine(party: StatementParty) {
    val details = listOfNotNull(party.phone, party.place)
    if (details.isNotEmpty()) {
        Text(
            details.joinToString(" · "),
            style = NocturneType.meta,
            color = Nocturne.neutral500,
            modifier = Modifier.padding(bottom = 8.dp)
        )
    }
}

@Composable
private fun Document(
    statement: Statement,
    currency: Currency,
    strings: Strings,
    onEditCreditNote: (com.stockbook.core.model.CreditNote) -> Unit = {},
    onEditPayment: (String) -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .card(Metrics.statRadius)
            .hairline(radius = Metrics.statRadius)
            .padding(16.dp)
    ) {
        Kicker(
            strings.dateSpan(
                from = strings.longDate(statement.range.start),
                to = strings.longDate(lastDay(statement.range))
            ),
            modifier = Modifier.padding(bottom = 12.dp)
        )

        Figure(strings.openingBalance, Money.text(statement.openingBalance, currency), muted = true)

        FadedRule(modifier = Modifier.padding(vertical = 10.dp))

        if (statement.isEmpty) {
            Text(
                strings.nothingInThisPeriod,
                style = NocturneType.meta,
                color = Nocturne.neutral500,
                modifier = Modifier.fillMaxWidth().padding(vertical = 14.dp)
            )
        } else {
            statement.entries.forEachIndexed { index, entry ->
                EntryRow(
                    entry = entry,
                    balance = statement.runningBalances[index],
                    currency = currency,
                    strings = strings,
                    onEditPayment = onEditPayment,
                    onEditCreditNote = onEditCreditNote
                )
                if (index < statement.entries.lastIndex) Spacer(Modifier.height(10.dp))
            }
        }

        FadedRule(modifier = Modifier.padding(vertical = 10.dp))

        Figure(chargedLabel(statement, strings), Money.text(statement.billed, currency))
        Spacer(Modifier.height(2.dp))
        Figure(settledLabel(statement, strings), Money.text(statement.received, currency))
        // Its own line, and only where there is one. Credit is not cash, and a
        // row saying "0.00 credited" on every statement teaches people to stop
        // reading the ones that are not zero.
        if (statement.credited > 0) {
            Spacer(Modifier.height(2.dp))
            Figure(strings.creditNotes, Money.text(statement.credited, currency))
        }

        FadedRule(modifier = Modifier.padding(vertical = 10.dp))

        // The number the whole document exists to state.
        Row(verticalAlignment = Alignment.Bottom, modifier = Modifier.fillMaxWidth()) {
            Text(
                strings.closingBalance,
                style = NocturneType.inter(13.0, FontWeight.Medium),
                color = Nocturne.text,
                modifier = Modifier.weight(1f)
            )
            Text(
                closingText(statement, currency, strings),
                style = NocturneType.bigNumber(22.0),
                color = if (statement.closingBalance > 0) Nocturne.accent else Nocturne.text
            )
        }
    }
}

@Composable
private fun Figure(label: String, value: String, muted: Boolean = false) {
    Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.fillMaxWidth()) {
        Text(
            label,
            style = if (muted) NocturneType.meta else NocturneType.inter(13.0),
            color = if (muted) Nocturne.neutral500 else Nocturne.text,
            modifier = Modifier.weight(1f)
        )
        Text(
            value,
            style = NocturneType.inter(13.0),
            color = if (muted) Nocturne.neutral500 else Nocturne.text
        )
    }
}

@Composable
private fun EntryRow(
    entry: Statement.Entry,
    balance: Double,
    currency: Currency,
    strings: Strings,
    onEditPayment: (String) -> Unit,
    onEditCreditNote: (com.stockbook.core.model.CreditNote) -> Unit = {}
) {
    // Either kind of payment can be deleted here; a bill or a delivery is changed
    // or taken out from inside the document itself, which is where the owner can
    // see what they are about to touch.
    val paymentId = when (entry) {
        is Statement.Entry.ForPayment -> entry.payment.id
        is Statement.Entry.ForSupplierPayment -> entry.payment.id
        else -> null
    }
    // Both of the things this screen can correct open the same way: tap the row,
    // get the sheet it was written on, with removal one button inside it. They
    // used to differ — a credit note opened, a payment armed a delete — which
    // meant the same gesture did two things depending on which row it landed on.
    //
    // A bill or a delivery is not corrected from here. Removing one puts stock
    // back on the shelf, and offering that from a row on a document somebody is
    // reading would be a second, worse route than the opened document itself.
    val creditNote = (entry as? Statement.Entry.ForCreditNote)?.note

    Column(
        modifier = Modifier
            .fillMaxWidth()
            .then(
                when {
                    creditNote != null -> Modifier.clickable { onEditCreditNote(creditNote) }
                    paymentId == null -> Modifier
                    else -> Modifier.clickable { onEditPayment(paymentId) }
                }
            )
    ) {
        Row(verticalAlignment = Alignment.Top, modifier = Modifier.fillMaxWidth()) {
            Column(modifier = Modifier.weight(1f)) {
                when (entry) {
                    is Statement.Entry.ForBill -> {
                        Text(
                            reference(entry, strings),
                            style = NocturneType.inter(13.0),
                            color = Nocturne.text
                        )
                        // Drawn only where there is something to say. A bill
                        // entered as a figure lists nothing, and an empty second
                        // line leaves a gap in a document somebody is checking
                        // line by line.
                        Detail(entry.bill.summary)
                    }
                    is Statement.Entry.ForPayment -> {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Glyph(Icon.confirm, size = 10.dp, tint = Nocturne.accent400)
                            Spacer(Modifier.width(5.dp))
                            Text(reference(entry, strings), style = NocturneType.inter(13.0), color = Nocturne.accent400)
                        }
                        entry.payment.note?.let {
                            Text(it, style = NocturneType.meta, color = Nocturne.neutral500)
                        }
                    }
                    is Statement.Entry.ForCreditNote -> {
                        Text(
                            reference(entry, strings),
                            style = NocturneType.inter(13.0),
                            color = Nocturne.accent400
                        )
                        // Why it was written, where the owner said. On a document
                        // somebody is checking against their own paper, "returned,
                        // damaged" is the difference between a figure they
                        // recognise and one they have to go and ask about.
                        Detail(entry.note.reason)
                    }
                    is Statement.Entry.ForPurchase -> {
                        Text(
                            reference(entry, strings),
                            style = NocturneType.inter(13.0),
                            color = Nocturne.text
                        )
                        // The delivery note's whole content on one line, and
                        // nothing at all where the supplier's bill named none.
                        Detail(entry.purchase.described)
                    }
                    is Statement.Entry.ForSupplierPayment -> {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Glyph(Icon.confirm, size = 10.dp, tint = Nocturne.accent400)
                            Spacer(Modifier.width(5.dp))
                            Text(reference(entry, strings), style = NocturneType.inter(13.0), color = Nocturne.accent400)
                        }
                        entry.payment.note?.let {
                            Text(it, style = NocturneType.meta, color = Nocturne.neutral500)
                        }
                    }
                    is Statement.Entry.ForTransfer -> {
                        // Named by the account at the other end — see
                        // `reference` — because there is no number to show. The
                        // reason is drawn under it for the same purpose a credit
                        // note's is: a figure the customer cannot place is one
                        // they have to come and ask about.
                        Text(
                            reference(entry, strings),
                            style = NocturneType.inter(13.0),
                            color = Nocturne.text
                        )
                        Detail(entry.transfer.note)
                    }
                }
                Text(strings.longDate(entry.date), style = NocturneType.meta, color = Nocturne.neutral500)
            }

            Spacer(Modifier.width(6.dp))

            Column(horizontalAlignment = Alignment.End) {
                Text(
                    amountText(entry, currency),
                    style = NocturneType.inter(13.0),
                    color = when (entry) {
                        is Statement.Entry.ForBill,
                        is Statement.Entry.ForPurchase -> Nocturne.text
                        is Statement.Entry.ForPayment,
                        is Statement.Entry.ForSupplierPayment,
                        is Statement.Entry.ForCreditNote -> Nocturne.accent400
                        // Whichever way this one moves the account, so it reads
                        // like the charge or the settlement it is.
                        is Statement.Entry.ForTransfer ->
                            if (entry.outgoing) Nocturne.accent400 else Nocturne.text
                    }
                )
                // The running balance beside every line: the column that turns a list
                // into a statement somebody can check.
                Text(Money.text(balance, currency), style = NocturneType.meta, color = Nocturne.neutral500)
            }
        }

    }
}

/**
 * The line under an entry's reference, where the entry has one.
 *
 * Null and blank are the same answer here — a bill with no lines on it and a
 * delivery with no product on it both have nothing to add — and both draw
 * nothing rather than an empty row the eye has to account for.
 */
@Composable
private fun Detail(text: String?) {
    if (text.isNullOrBlank()) return
    Text(
        text,
        style = NocturneType.meta,
        color = Nocturne.neutral500,
        maxLines = 2,
        overflow = TextOverflow.Ellipsis
    )
}

/**
 * What a row is called on screen — the printed statement's own rule, borrowed
 * rather than restated.
 *
 * The two are read side by side when somebody checks a PDF against the app, and
 * a row named two different ways is a row they reconcile by eye.
 */
private fun reference(entry: Statement.Entry, strings: Strings): String =
    StatementDocument.reference(entry, strings)

private fun amountText(entry: Statement.Entry, currency: Currency): String = when (entry) {
    is Statement.Entry.ForBill -> Money.text(entry.bill.total, currency)
    is Statement.Entry.ForPurchase -> Money.text(entry.purchase.total, currency)
    // A minus sign on both kinds of payment: it is what the account moves by, and
    // on a supplier's statement that is money leaving rather than arriving.
    is Statement.Entry.ForCreditNote -> "− ${Money.text(entry.note.total, currency)}"
    is Statement.Entry.ForPayment -> "− ${Money.text(entry.payment.amount, currency)}"
    is Statement.Entry.ForSupplierPayment -> "− ${Money.text(entry.payment.amount, currency)}"
    // The sign says which way this account moved, not which way the money went —
    // no money went anywhere. Out of this account is a reduction like a payment;
    // into it is a charge like a bill.
    is Statement.Entry.ForTransfer ->
        if (entry.outgoing) "− ${Money.text(entry.transfer.amount, currency)}"
        else Money.text(entry.transfer.amount, currency)
}

/**
 * "Billed" and "Received" are the customer's words. On a supplier's account they
 * read backwards — the shop is the one being billed — so the two figures are
 * named by which way the account runs.
 */
private fun chargedLabel(statement: Statement, strings: Strings): String =
    if (statement.party.isSupplier) strings.purchasedInPeriod else strings.billedInPeriod

private fun settledLabel(statement: Statement, strings: Strings): String =
    if (statement.party.isSupplier) strings.paidOutInPeriod else strings.receivedInPeriod

private fun closingText(statement: Statement, currency: Currency, strings: Strings): String =
    if (statement.closingBalance < 0) {
        strings.inAdvance(Money.text(-statement.closingBalance, currency))
    } else {
        Money.text(statement.closingBalance, currency)
    }

/**
 * The last day this statement can honestly say it covers — the range's own rule,
 * so the screen and the PDF are headed with the same date.
 */
private fun lastDay(range: StatementRange): Instant = range.asOf()
